plugins/modules/keycloak_authz_permission.py

```diff
@@ -0,0 +1,471 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2017, Eike Frost <ei@kefro.st>
+# Copyright (c) 2021, Christophe Gilles <christophe.gilles54@gmail.com>
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or
+# https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_authz_permission
+
+short_description: Allows administration of Keycloak client authorization permissions via Keycloak API.
+
+description:
+    - This module allows the administration of Keycloak client authorization permissions via the Keycloak REST
+      API. Authorization permissions are only available if a client has Authorization enabled.
+
+    - There are some peculiarities in JSON paths and payloads for authorization permissions. In particular
+      POST and PUT operations are targeted at permission endpoints, whereas GET requests go to policies
+      endpoint. To make matters more interesting the JSON responses from GET requests return data in a
+      different format than what is expected for POST and PUT. The end result is that it is not possible to
+      detect changes to things like policies, scopes or resources - at least not without a large number of
+      additional API calls. Therefore this module always updates authorization permissions instead of
+      attempting to determine if changes are truly needed.
+
+    - This module requires access to the REST API via OpenID Connect; the user connecting and the realm
+      being used must have the requisite access rights. In a default Keycloak installation, admin-cli
+      and an admin user would work, as would a separate realm definition with the scope tailored
+      to your needs and a user having the expected roles.
+
+    - The names of module options are snake_cased versions of the camelCase options used by Keycloak.
+      The Authorization Services paths and payloads have not officially been documented by the Keycloak project.
+      https://www.puppeteers.net/blog/keycloak-authorization-services-rest-api-paths-and-payload/
+
+attributes:
+    check_mode:
+        support: none
+    diff_mode:
+        support: none
```

Comment:
All other keycloak modules support check mode, and almost all others also support diff mode. Is there a technical reason why not at least check mode can be implemented here?

Reply:
I can implement check_mode for create and remove operations. What I cannot do is check if changes are required to the desired state of already existing objects. The full explanation is in a comment above.

I guess the discussion revolves around what is the correct value for *result['changed']* when you cannot know for sure if something changed (or would change).

Reply:
Check mode with the same behavior as the real implementation (i.e. always return `changed=True` if you're not sure) is already good enough IMO.

Reply:
Thanks for the feedback @felixfontein. I'll make the necessary changes.

Reply:
Fixed the `result['changed']` issue and added support for check_mode. Info module next.

Reply:
I implemented the "query_state" as an info module and ripped away that functionality from the main module. I also made the integration tests use the info module.